### PR TITLE
feat(plasma-hope): draggable scroll in carousel

### PR DESCRIPTION
**Pull request by kayman233**
Добавлена возможность прокрутки `Carousel` с помощью мышки

ezgif-3-888a6fcbac

:baby_chick: Download canary assets:

[colors--canary.476.4686853378.xml](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/colors--canary.476.4686853378.xml)  
[color_sberHealth_ios-swift--canary.476.4686853378.swift](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/color_sberHealth_ios-swift--canary.476.4686853378.swift)  
[color_sberHealth_kotlin--canary.476.4686853378.kt](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/color_sberHealth_kotlin--canary.476.4686853378.kt)  
[color_sberHealth_react-native--canary.476.4686853378.ts](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/color_sberHealth_react-native--canary.476.4686853378.ts)  
[color_sbermarket_business_ios-swift--canary.476.4686853378.swift](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/color_sbermarket_business_ios-swift--canary.476.4686853378.swift)  
[color_sbermarket_business_kotlin--canary.476.4686853378.kt](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/color_sbermarket_business_kotlin--canary.476.4686853378.kt)  
[color_sbermarket_business_react-native--canary.476.4686853378.ts](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/color_sbermarket_business_react-native--canary.476.4686853378.ts)  
[color_sbermarket_ios-swift--canary.476.4686853378.swift](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/color_sbermarket_ios-swift--canary.476.4686853378.swift)  
[color_sbermarket_kotlin--canary.476.4686853378.kt](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/color_sbermarket_kotlin--canary.476.4686853378.kt)  
[color_sbermarket_metro_ios-swift--canary.476.4686853378.swift](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/color_sbermarket_metro_ios-swift--canary.476.4686853378.swift)  
[color_sbermarket_metro_kotlin--canary.476.4686853378.kt](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/color_sbermarket_metro_kotlin--canary.476.4686853378.kt)  
[color_sbermarket_metro_react-native--canary.476.4686853378.ts](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/color_sbermarket_metro_react-native--canary.476.4686853378.ts)  
[color_sbermarket_react-native--canary.476.4686853378.ts](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/color_sbermarket_react-native--canary.476.4686853378.ts)  
[color_sbermarket_selgros_ios-swift--canary.476.4686853378.swift](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/color_sbermarket_selgros_ios-swift--canary.476.4686853378.swift)  
[color_sbermarket_selgros_kotlin--canary.476.4686853378.kt](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/color_sbermarket_selgros_kotlin--canary.476.4686853378.kt)  
[color_sbermarket_selgros_react-native--canary.476.4686853378.ts](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/color_sbermarket_selgros_react-native--canary.476.4686853378.ts)  
[color_sbermarket_vinlab_ios-swift--canary.476.4686853378.swift](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/color_sbermarket_vinlab_ios-swift--canary.476.4686853378.swift)  
[color_sbermarket_vinlab_kotlin--canary.476.4686853378.kt](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/color_sbermarket_vinlab_kotlin--canary.476.4686853378.kt)  
[color_sbermarket_vinlab_react-native--canary.476.4686853378.ts](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/color_sbermarket_vinlab_react-native--canary.476.4686853378.ts)  
[color_sberprime_ios-swift--canary.476.4686853378.swift](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/color_sberprime_ios-swift--canary.476.4686853378.swift)  
[color_sberprime_kotlin--canary.476.4686853378.kt](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/color_sberprime_kotlin--canary.476.4686853378.kt)  
[color_sberprime_react-native--canary.476.4686853378.ts](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/color_sberprime_react-native--canary.476.4686853378.ts)  
[PlasmaTokensColor--canary.476.4686853378.swift](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/PlasmaTokensColor--canary.476.4686853378.swift)  
[shadow_sbermarket_react-native--canary.476.4686853378.ts](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/shadow_sbermarket_react-native--canary.476.4686853378.ts)  
[typo_mage_ios-swift--canary.476.4686853378.swift](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/typo_mage_ios-swift--canary.476.4686853378.swift)  
[typo_mage_kotlin--canary.476.4686853378.kt](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/typo_mage_kotlin--canary.476.4686853378.kt)  
[typo_mage_react-native--canary.476.4686853378.ts](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/typo_mage_react-native--canary.476.4686853378.ts)  
[typo_plasma_ios-swift--canary.476.4686853378.swift](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/typo_plasma_ios-swift--canary.476.4686853378.swift)  
[typo_plasma_kotlin--canary.476.4686853378.kt](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/typo_plasma_kotlin--canary.476.4686853378.kt)  
[typo_plasma_react-native--canary.476.4686853378.ts](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/typo_plasma_react-native--canary.476.4686853378.ts)  
[typo_ruler_ios-swift--canary.476.4686853378.swift](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/typo_ruler_ios-swift--canary.476.4686853378.swift)  
[typo_ruler_kotlin--canary.476.4686853378.kt](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/typo_ruler_kotlin--canary.476.4686853378.kt)  
[typo_ruler_react-native--canary.476.4686853378.ts](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/typo_ruler_react-native--canary.476.4686853378.ts)  
[typo_sage_ios-swift--canary.476.4686853378.swift](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/typo_sage_ios-swift--canary.476.4686853378.swift)  
[typo_sage_kotlin--canary.476.4686853378.kt](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/typo_sage_kotlin--canary.476.4686853378.kt)  
[typo_sage_react-native--canary.476.4686853378.ts](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/typo_sage_react-native--canary.476.4686853378.ts)  
[typo_sbermarket_ios-swift--canary.476.4686853378.swift](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/typo_sbermarket_ios-swift--canary.476.4686853378.swift)  
[typo_sbermarket_kotlin--canary.476.4686853378.kt](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/typo_sbermarket_kotlin--canary.476.4686853378.kt)  
[typo_sbermarket_react-native--canary.476.4686853378.ts](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/typo_sbermarket_react-native--canary.476.4686853378.ts)  
[typo_soulmate_ios-swift--canary.476.4686853378.swift](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/typo_soulmate_ios-swift--canary.476.4686853378.swift)  
[typo_soulmate_kotlin--canary.476.4686853378.kt](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/typo_soulmate_kotlin--canary.476.4686853378.kt)  
[typo_soulmate_react-native--canary.476.4686853378.ts](https://github.com/salute-developers/plasma/releases/download/v0.0.0-canary/typo_soulmate_react-native--canary.476.4686853378.ts)

<details>
  <summary>📦 Published PR as canary version: <code>Canary Versions</code></summary>
  <br />

  :sparkles: Test out this PR locally via:
  
  ```bash
  npm install @salutejs/plasma-b2c@1.175.0-canary.476.4686853378.0
  npm install @salutejs/plasma-hope@0.33.0-canary.476.4686853378.0
  npm install @salutejs/plasma-web@1.200.0-canary.476.4686853378.0
  # or 
  yarn add @salutejs/plasma-b2c@1.175.0-canary.476.4686853378.0
  yarn add @salutejs/plasma-hope@0.33.0-canary.476.4686853378.0
  yarn add @salutejs/plasma-web@1.200.0-canary.476.4686853378.0
  ```
</details>

### CODE REVIEW

**Comment by Salute-Eva** [comment]
Theme Builder app deployed!
  
  http://plasma.sberdevices.ru/pr/plasma-theme-builder-pr-476/

**Comment by Salute-Eva** [comment]
**Component performance testing**

Result: **🟢 Success**

Check out report in job artifacts!

**Comment by Salute-Eva** [comment]
Demo canvas app deployed!
  
  http://plasma.sberdevices.ru/pr/demo-canvas-app-pr-476/

**Comment by Salute-Eva** [comment]
Documentation preview deployed!
  
  website: http://plasma.sberdevices.ru/pr/pr-476/
  ui storybook: http://plasma.sberdevices.ru/pr/pr-476/ui-storybook/
  web storybook: http://plasma.sberdevices.ru/pr/pr-476/web-storybook/
  b2c storybook: http://plasma.sberdevices.ru/pr/pr-476/b2c-storybook/
  temple storybook: http://plasma.sberdevices.ru/pr/pr-476/temple-storybook/

**Comment by Salute-Eva** [comment]
**Component performance testing**

Result: **🟢 Success**

Check out report in job artifacts!

**Comment by Salute-Eva** [comment]
Theme Builder app deployed!
  
  http://plasma.sberdevices.ru/pr/plasma-theme-builder-pr-476/

**Comment by Salute-Eva** [comment]
Documentation preview deployed!
  
  website: http://plasma.sberdevices.ru/pr/pr-476/
  ui storybook: http://plasma.sberdevices.ru/pr/pr-476/ui-storybook/
  web storybook: http://plasma.sberdevices.ru/pr/pr-476/web-storybook/
  b2c storybook: http://plasma.sberdevices.ru/pr/pr-476/b2c-storybook/
  temple storybook: http://plasma.sberdevices.ru/pr/pr-476/temple-storybook/

**Comment by Salute-Eva** [comment]
**Component performance testing**

Result: **🟢 Success**

Check out report in job artifacts!

**Comment by Yakutoc** [comment]
@kayman233 

Допиши пожалуйста, что и почему делает этот PR.

Если получиться гифку сделать с тем как стало вообще будет супер )

File: packages/plasma-hope/src/components/Carousel/hooks.tsx [hunk 1]
```diff
@@ -0,0 +1,61 @@
+import { useCallback, useEffect, useState } from 'react';
+
+const SCROLL_SPEED = 2;
```
[the remark this answers is not in the document]
Comment (kayman233) [reply packages/plasma-hope/src/components/Carousel/hooks.tsx:3]: Чисто эмпирически оказалось самым адекватным в плане скорости и плавности движений

File: packages/plasma-hope/src/components/Carousel/Carousel.tsx [hunk 2]
```diff
@@ -47,6 +49,8 @@ export const Carousel = React.forwardRef<HTMLDivElement, CarouselProps>(function
     });
     const handleRef = useForkRef(scrollRef, ref);
 
+    useDragScroll(scrollRef);
```
[the remark this answers is not in the document]
Comment (kayman233) [reply packages/plasma-hope/src/components/Carousel/Carousel.tsx:52]: Гена говорил, что по умолчанию хотим, если я правильно понял

[the remark this answers is not in the document]
Comment (Yeti-or) [reply packages/plasma-hope/src/components/Carousel/Carousel.tsx:52]: ну мы можем конечно это на мобилке отрывать, но чет не знаю

File: packages/plasma-hope/src/components/Carousel/hooks.tsx [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+            }
+        },
+        [scrollRef, setDragging, setStartX, setScrollLeft, handleMouseUp, handleMouseMove],
+    );
+
+    useEffect(() => {
+        if (scrollRef && scrollRef.current) {
+            scrollRef.current.addEventListener('mousedown', handleMouseDown);
+            scrollRef.current.addEventListener('mouseup', handleMouseUp);
+            scrollRef.current.addEventListener('mouseleave', handleMouseUp);
```
Comment (neretin-trike) [inline packages/plasma-hope/src/components/Carousel/hooks.tsx:48]: а и здесь и должно быть handleMouseUp?

**Review by Yeti-or — COMMENTED** [review]
Мне кажется нужно навесить в таком случае user-select: none
потому что раньше текст на карточках можно было выделять,  а теперь можно только в Safari и это криво смотрится

File: packages/plasma-hope/src/components/Carousel/Carousel.tsx (hunk 2 above)
[the remark this answers is not in the document]
Comment (neretin-trike) [reply packages/plasma-hope/src/components/Carousel/Carousel.tsx:52]: кажется, что можно всегда включенным оставлять, пока нет кейсов, при которых понадобится явно вырубать

**Review by neretin-trike — APPROVED** [review]

**Review by Yakutoc — APPROVED** [review]

**Comment by Salute-Eva** [comment]
Demo canvas app deployed!
  
  http://plasma.sberdevices.ru/pr/demo-canvas-app-pr-476/

**Comment by Salute-Eva** [comment]
Documentation preview deployed!
  
  website: http://plasma.sberdevices.ru/pr/pr-476/
  ui storybook: http://plasma.sberdevices.ru/pr/pr-476/ui-storybook/
  web storybook: http://plasma.sberdevices.ru/pr/pr-476/web-storybook/
  b2c storybook: http://plasma.sberdevices.ru/pr/pr-476/b2c-storybook/
  temple storybook: http://plasma.sberdevices.ru/pr/pr-476/temple-storybook/

**Comment by Salute-Eva** [comment]
**Component performance testing**

Result: **🟢 Success**

Check out report in job artifacts!

**Pull request merged**